# Read file into Dataframes

In [69]:
import numpy as np 
import pandas as pd 
user = pd.read_excel('RnAPortal.xlsx', 'PORTAL_USERS')
user.head()

USER_PID  FIRST_NAME LAST_NAME     DEPARTMENT_NAME     DIVISION  \
0    230168  Raphaelito     Anjos              Grêmio  Sporting CP   
1    203279  Wellington       Nem    Shakhtar Donetsk   SL Benfica   
2    209722  Marquinhos   Pedroso           FC Dallas  Sporting CP   
3    244497  Marquinhos  Cipriano    Shakhtar Donetsk  Sporting CP   
4    231394  Hildeberto   Pereira  Vitória de Setúbal   SL Benfica   

  EMPLOYEE_CLASS GRADE_SHORT        COUNTRY       CITY  
0            PER          VP  UNITED STATES   New York  
1            PER          VP  UNITED STATES   New York  
2            PER          VP          INDIA       Pune  
3            PER         AVP      SINGAPORE  Singapore  
4            PER         MDR    SWITZERLAND      Zrich

In [70]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3408 entries, 0 to 3407
Data columns (total 9 columns):
USER_PID           3408 non-null int64
FIRST_NAME         3408 non-null object
LAST_NAME          3408 non-null object
DEPARTMENT_NAME    3408 non-null object
DIVISION           3408 non-null object
EMPLOYEE_CLASS     3374 non-null object
GRADE_SHORT        2095 non-null object
COUNTRY            3408 non-null object
CITY               3399 non-null object
dtypes: int64(1), object(8)
memory usage: 239.7+ KB


<font color="##0000dd" size="3">**User**: 3408 users, missing values in EMPLOYEE_CLASS, GRADE_SHORT, COUNTRY and CITY</font>

In [66]:
app = pd.read_excel('RnAPortal.xlsx', 'PORTAL_APPS')
app.head()

APP_ID                 APP_NAME  BUSINESS_DOMAIN             CATEGORY  \
0     403    Sketch - Draw & Paint      CREDIT RISK       ART_AND_DESIGN   
1     301  Wattpad 📖 Free Books  PRODUCT CONTROL  BOOKS_AND_REFERENCE   
2     163            Amazon Kindle  PRODUCT CONTROL  BOOKS_AND_REFERENCE   
3     627        Google Play Books             RFIT  BOOKS_AND_REFERENCE   
4     677        Indeed Job Search             RFIT             BUSINESS   

          APP_TYPE  AVG_RATING  
0  WEB APPLICATION         NaN  
1  WEB APPLICATION         NaN  
2  WEB APPLICATION         NaN  
3          TABLEAU         5.0  
4          TABLEAU         NaN

In [67]:
app.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 6 columns):
APP_ID             262 non-null int64
APP_NAME           262 non-null object
BUSINESS_DOMAIN    262 non-null object
CATEGORY           262 non-null object
APP_TYPE           262 non-null object
AVG_RATING         28 non-null float64
dtypes: float64(1), int64(1), object(4)
memory usage: 12.4+ KB


<font color="##0000dd" size="3">**App**: 262 applications, missing values in AVG_RATING</font>

In [388]:
# Getting number of users and apps from the dataset.
user_ids = user.USER_PID.unique().tolist()
app_ids = app.APP_ID.unique().tolist()
print('Number of users: {}'.format(len(user_ids)))
print('Number of apps: {}'.format(len(app_ids)))

Number of users: 3408
Number of apps: 262


<font color="##0000dd" size="3">Number of users: 3408</font><br>
<font color="##0000dd" size="3">Number of apps: 262</font><br>

In [95]:
ent = pd.read_excel('RnAPortal.xlsx', 'APP_ENTITLEMENTS')
ent.head()

APP_ID  USER_PID
0       5    238089
1       5    167431
2       5    244497
3       5    242629
4       5    194728

In [96]:
ent.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84848 entries, 0 to 84847
Data columns (total 2 columns):
APP_ID      84848 non-null int64
USER_PID    84848 non-null int64
dtypes: int64(2)
memory usage: 1.3 MB


<font color="##0000dd" size="3">**Entitlement**: no missing values</font>

In [288]:
wspace = pd.read_excel('RnAPortal.xlsx', 'USER_WORKSPACE_APPS')
wspace.head()

USER_PID  USER_WORKSPACE_ID APP_ID
0    210450                802      2
1    210450                802      3
2    210450                802    431
3    220925                803      2
4    220925                803      3

In [289]:
wspace.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3698 entries, 0 to 3697
Data columns (total 3 columns):
USER_PID             3698 non-null int64
USER_WORKSPACE_ID    3698 non-null int64
APP_ID               3698 non-null object
dtypes: int64(2), object(1)
memory usage: 86.8+ KB


<font color="##0000dd" size="3">**Workspace**: no missing values, but APP_ID is string</font>

In [290]:
# show the row where APP_ID is string
wspace[wspace['APP_ID'].apply(lambda x: isinstance(x, str))]

USER_PID  USER_WORKSPACE_ID                  APP_ID
3224    142707               2136   UK IWM Legal Entities

In [365]:
# delete the row where APP_ID is string
wspace = wspace[wspace['APP_ID'].apply(lambda x: isinstance(x, int))]
wspace['APP_ID'] = wspace['APP_ID'].astype(int)
wspace.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3697 entries, 0 to 3697
Data columns (total 3 columns):
USER_PID             3697 non-null int64
USER_WORKSPACE_ID    3697 non-null int64
APP_ID               3697 non-null int64
dtypes: int64(3)
memory usage: 115.5 KB


In [375]:
print('Number of users who created wspace is:', wspace['USER_PID'].nunique())
print('Number of apps appearing in work space is:', wspace['APP_ID'].nunique())
wspace_user = wspace.USER_PID.unique().tolist()
wspace_app = wspace.APP_ID.unique().tolist()

Number of users who created wspace is: 1041
Number of apps appearing in work space is: 202


<font color="##0000dd" size="3">Number of users who created wspace is: 1041</font><br>
<font color="##0000dd" size="3">Number of apps appearing in work space is: 202</font><br>

In [281]:
rating = pd.read_excel('RnAPortal.xlsx', 'APP_RATINGS')
rating.head()

USER_PID  APP_ID  OVERALL_RATING  DATA_RATING  UI_RATING  USABILITY_RATING  \
0    205995     165               1          NaN        NaN               NaN   
1    171579     121               5          NaN        NaN               NaN   
2    171189     278               4          NaN        NaN               NaN   
3    209307     382               5          NaN        NaN               NaN   
4    187275     121               1          NaN        NaN               NaN   

                                            COMMENTS  
0                                                NaN  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4  I was able to login and upload commentaries on...

In [24]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 7 columns):
USER_PID            54 non-null int64
APP_ID              54 non-null int64
OVERALL_RATING      54 non-null int64
DATA_RATING         16 non-null float64
UI_RATING           16 non-null float64
USABILITY_RATING    16 non-null float64
COMMENTS            30 non-null object
dtypes: float64(3), int64(3), object(1)
memory usage: 3.0+ KB


<font color="##0000dd" size="3">**Rating**: data sparsity, few data</font>

In [378]:
print('Number of users who rated apps is:', rating['USER_PID'].nunique())
print('Number of apps being rated is:', rating['APP_ID'].nunique())
rating_user = rating.USER_PID.unique().tolist()
rating_app = rating.APP_ID.unique().tolist()

Number of users who rated apps is: 36
Number of apps being rated is: 27


<font color="##0000dd" size="3">Number of users who rated apps is: 36</font><br>
<font color="##0000dd" size="3">Number of apps being rated is: 27</font><br>

In [353]:
log = pd.read_excel('RnAPortal.xlsx', 'APP_ACCESS_LOG')
log.head()

APP_ID  USER_PID          START_TIME            END_TIME  USAGE_TIME_MINS
0    82.0    228600 2018-03-27 15:20:53 2018-03-27 15:24:22              3.5
1    82.0    228600 2018-03-27 15:26:39 2018-03-27 15:28:18              1.7
2     1.0    209782 2018-03-27 16:46:55 2018-03-27 16:49:28              2.6
3    82.0    228600 2018-03-28 14:47:02 2018-03-28 14:48:45              1.7
4     2.0    198118 2018-03-28 14:51:14 2018-03-28 14:55:27              4.2

In [354]:
log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 217219 entries, 0 to 217218
Data columns (total 5 columns):
APP_ID             217216 non-null float64
USER_PID           217219 non-null int64
START_TIME         217219 non-null datetime64[ns]
END_TIME           217219 non-null datetime64[ns]
USAGE_TIME_MINS    217219 non-null float64
dtypes: datetime64[ns](2), float64(2), int64(1)
memory usage: 8.3 MB


<font color="##0000dd" size="3">**Log**: missing APP_ID, need pivoting</font>

In [367]:
# delete rows with missing APP_ID
log = log.dropna()
log['APP_ID'] = log['APP_ID'].astype(int)
log.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 217216 entries, 0 to 217218
Data columns (total 5 columns):
APP_ID             217216 non-null int64
USER_PID           217216 non-null int64
START_TIME         217216 non-null datetime64[ns]
END_TIME           217216 non-null datetime64[ns]
USAGE_TIME_MINS    217216 non-null float64
dtypes: datetime64[ns](2), float64(1), int64(2)
memory usage: 9.9 MB


In [387]:
print('Number of users who interacted with apps is:', log['USER_PID'].nunique())
print('Number of apps being used is:', log['APP_ID'].nunique())
log_user = log.USER_PID.unique().tolist()
log_app = log.APP_ID.unique().tolist()

Number of users who interacted with apps is: 1541
Number of apps being used is: 211


<font color="##0000dd" size="3">Number of users who interacted with apps is: 1541</font><br>
<font color="##0000dd" size="3">Number of apps being used is: 211</font><br>

In [385]:
user_with_data = set(wspace_user).union(set(rating_user)).union(set(log_user))
print('The union of user with data is:', len(user_with_data))
app_with_data = set(wspace_app).union(set(rating_app)).union(set(log_app))
print('The union of app with data is:', len(app_with_data))

The union of user with data is: 1634
The union of app with data is: 224


<font color="##0000dd" size="3">
    
Observation:

1. Total number of users: 3408, total number of apps: 262.<br>
2. Number of users who interacted with apps is: 1541. Number of apps being used is: 211. (based on log data) <br>
3. Number of users of any kind of record filtering is 1634, number of apps of any kind of record is 224.

Temporary Conlusion: 

1. It is noticeable that some users rate or create workspace without actually using apps. Can this data be counted as reliable? <br>
2. We can use workspace, log, rating, app category and user attributes for collaborative filtering recommendation. Rating and log can be used to rank recommendations.<br>
3. For new users, use collaborative filtering recommendation based on user attributes.</font>
4. For existing users, use collaborative filtering 

# Combine Dataframes

In [374]:
# this shows how user interact with apps
user_log = pd.merge(user, log, on='USER_PID')
user_log.head()

USER_PID  FIRST_NAME LAST_NAME   DEPARTMENT_NAME     DIVISION  \
0    203279  Wellington       Nem  Shakhtar Donetsk   SL Benfica   
1    203279  Wellington       Nem  Shakhtar Donetsk   SL Benfica   
2    203279  Wellington       Nem  Shakhtar Donetsk   SL Benfica   
3    203279  Wellington       Nem  Shakhtar Donetsk   SL Benfica   
4    209722  Marquinhos   Pedroso         FC Dallas  Sporting CP   

  EMPLOYEE_CLASS GRADE_SHORT        COUNTRY      CITY  APP_ID  \
0            PER          VP  UNITED STATES  New York     321   
1            PER          VP  UNITED STATES  New York     321   
2            PER          VP  UNITED STATES  New York     321   
3            PER          VP  UNITED STATES  New York     321   
4            PER          VP          INDIA      Pune     103   

           START_TIME            END_TIME  USAGE_TIME_MINS  
0 2019-01-18 19:30:44 2019-01-18 20:00:43             30.0  
1 2019-01-18 20:02:12 2019-01-18 20:50:12             48.0  
2 2019-01-18 19:00:19 2019-01-18 19:30:19             30.0  
3 2019-01-18 18:00:22 2019-01-18 18:59:22             59.0  
4 2018-04-13 02:58:27 2018-04-13 03:41:21             42.9

In [371]:
# user & mean_usage_time matrix
user_log.pivot_table(index='USER_PID',columns='APP_ID',values='USAGE_TIME_MINS').head()

APP_ID    1    2    3    4    5    12   14   16   21   61  ...   827  845  \
USER_PID                                                   ...              
1179      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
14907     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
20801     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
24630     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
47201     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   

APP_ID    866  867  885  886  887  888  905  906  
USER_PID                                          
1179      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
14907     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
20801     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
24630     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
47201     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 211 columns]

In [372]:
# this shows how user rate apps
user_rating = pd.merge(user, rating, on='USER_PID')
user_rating.head()

USER_PID FIRST_NAME  LAST_NAME DEPARTMENT_NAME    DIVISION EMPLOYEE_CLASS  \
0    230461  Leordinho       Paes          Santos  SL Benfica            PER   
1    201883  Alejandro  Rodríguez          Empoli  SL Benfica            PER   
2    200315   Cristian      Tello      Real Betis  SL Benfica            PER   
3    230328   Pierrino      Faria          Santos  SL Benfica            PER   
4    230387   Heitinho      Zanon           Bahia  SL Benfica            PER   

  GRADE_SHORT         COUNTRY       CITY  APP_ID  OVERALL_RATING  DATA_RATING  \
0          VP     SWITZERLAND      Zrich     104               1          NaN   
1          VP          POLAND    Wroclaw     607               4          5.0   
2         DIR     SWITZERLAND      Zrich     445               5          5.0   
3         DIR  UNITED KINGDOM     London     121               1          NaN   
4         AVP          BRAZIL  Sao Paulo     278               2          3.0   

   UI_RATING  USABILITY_RATING  \
0        NaN               NaN   
1        4.0               5.0   
2        5.0               5.0   
3        NaN               NaN   
4        2.0               1.0   

                                            COMMENTS  
0                                                NaN  
1       Great for owners to track application usage.  
2                                                NaN  
3  The provision review tab isn't particularly go...  
4  System hard to filter and not able to share, d...

In [373]:
# user & rating matrix
user_rating.pivot_table(index='USER_PID',columns='APP_ID',values='OVERALL_RATING').head()

APP_ID    2    101  104  121  165  190  242  278  281  321 ...   625  626  \
USER_PID                                                   ...              
138981    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
171189    NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0  NaN  NaN ...   NaN  NaN   
171579    NaN  NaN  NaN  5.0  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   
184819    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  5.0 ...   NaN  NaN   
187275    NaN  NaN  NaN  1.0  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN   

APP_ID    627  665  666  667  679  683  705  706  
USER_PID                                          
138981    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
171189    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
171579    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
184819    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
187275    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 27 columns]

In [300]:
# this shows app category and user's ways to organize apps
wspace_app = pd.merge(app, wspace, on='APP_ID', how='left')
wspace_app.head()

APP_ID                 APP_NAME  BUSINESS_DOMAIN             CATEGORY  \
0     403    Sketch - Draw & Paint      CREDIT RISK       ART_AND_DESIGN   
1     403    Sketch - Draw & Paint      CREDIT RISK       ART_AND_DESIGN   
2     301  Wattpad 📖 Free Books  PRODUCT CONTROL  BOOKS_AND_REFERENCE   
3     301  Wattpad 📖 Free Books  PRODUCT CONTROL  BOOKS_AND_REFERENCE   
4     301  Wattpad 📖 Free Books  PRODUCT CONTROL  BOOKS_AND_REFERENCE   

          APP_TYPE  AVG_RATING  USER_PID  USER_WORKSPACE_ID  
0  WEB APPLICATION         NaN  225647.0             1557.0  
1  WEB APPLICATION         NaN  172018.0             2599.0  
2  WEB APPLICATION         NaN  184200.0             1041.0  
3  WEB APPLICATION         NaN  225261.0             1133.0  
4  WEB APPLICATION         NaN  224546.0             1110.0